config/src/network_id.rs

```diff
@@ -152,6 +152,64 @@ impl NetworkId {
         matches!(self, NetworkId::Validator)
     }
 
+    pub fn upstream_roles(&self, role: &RoleType) -> &'static [PeerRole] {
+        match self {
+            NetworkId::Validator => match role {
+                RoleType::Validator => &[PeerRole::Validator],
+                RoleType::FullNode => &[],
+            },
+            NetworkId::Public => &[
+                PeerRole::PreferredUpstream,
+                PeerRole::Upstream,
+                PeerRole::ValidatorFullNode,
+            ],
+            NetworkId::Private(_) => {
+                if self.is_vfn_network() {
+                    match role {
+                        RoleType::Validator => &[],
+                        RoleType::FullNode => &[PeerRole::Validator, PeerRole::ValidatorFullNode],
+                    }
+                } else {
+                    // TODO: This is not a well defined network
+                    &[]
+                }
+            }
+        }
+    }
+
+    pub fn downstream_roles(&self, role: &RoleType) -> &'static [PeerRole] {
+        match self {
+            NetworkId::Validator => match role {
+                RoleType::Validator => &[PeerRole::Validator],
+                RoleType::FullNode => &[],
+            },
+            // In order to allow fallbacks, we must allow for nodes to accept Validators and ValidatorFullNodes
+            NetworkId::Public => &[
+                PeerRole::Downstream,
+                PeerRole::Validator,
+                PeerRole::ValidatorFullNode,
+                PeerRole::Known,
+                PeerRole::Unknown,
+            ],
```

Comment:
This seems weird, a validator only connects to other validators, that to me is a peer, same as VFN connecting to another VFN, downstream is something different afaiu.... also a validator would never connect to a public network.

Reply:
So, yes it's a `peer` not really upstream or downstream for Validator.  This is a lot of the problematic issues of our network being P2P, but attaching a Client/Server paradigm on top.  I'm not quite sure where there will be gaps just because one place works and another doesn't.